rust/src/delta_datafusion.rs

```diff
@@ -263,29 +338,70 @@ impl TableProvider for delta::DeltaTable {
             self.object_store(),
         );
 
-        // TODO prune files based on file statistics and filter expressions
-        let partitions = self
-            .get_state()
-            .files()
-            .iter()
-            .map(|action| {
-                Ok(vec![PartitionedFile::new(
-                    action.path.clone(),
-                    action.size as u64,
-                )])
-            })
-            .collect::<DataFusionResult<_>>()?;
+        // TODO we group files together by their partition values. If the table is partitioned
+        // and partitions are somewhat evenly distributed, probably not the worst choice ...
+        // However we may want to do some additional balancing in case we are far off from the above.
+        let mut file_groups: HashMap<Vec<ScalarValue>, Vec<PartitionedFile>> = HashMap::new();
+        if !filters.is_empty() {
+            combine_filters(filters)
+                .map::<DataFusionResult<_>, _>(|predicate| {
+                    let pruning_predicate = PruningPredicate::try_new(predicate, schema.clone())?;
+                    let files_to_prune = pruning_predicate.prune(self)?;
```

Comment:
I was a bit concerned about interpreting the filters properly, but luckily the hard part was done 😆.